### ISSUE TITLE
[Broken Build]: Build error with special characters in username on Windows

### ISSUE THREAD

**Issue by Danaozhong** [dotnet/msbuild#9528]
### Issue Description

I am facing an build error after updating Visual Studio 2022 and installing the latest .NET framework. Every CMake build project fails with an error message from `MSBuild`. The issue seems to be related to the way `MSBuild` uses temporary paths.
I am using CMake, and was able to reproduce the issue with any CMake project I try to build. Before updating MSVC, these projects could be built without issues.

See the following build output:
```powershell
PS C:\Work\CPP\cpp-cmake-bare\bin> cmake --build .
MSBuild version 17.8.3+195e7f5a3 for .NET Framework

  'C:\Users\123\AppData\Local\Temp\MSBuildTempFirstNameLastName' is not recognized as an internal or external command,
  operable program or batch file.
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(237,5): e
rror MSB8066: Custom build for 'C:\Work\CPP\cpp-cmake-bare\bin\CMakeFiles\d0c3deefe69ea22f50ff94701eaf2da5\gene
rate.stamp.rule' exited with code 1. [C:\Work\CPP\cpp-cmake-bare\bin\ZERO_CHECK.vcxproj]
```
Looking at my temporary folder, I can see that the folder `C:\Users\123\AppData\Local\Temp\MSBuildTempFirstNameLastName` indeed does not exist, but the folder `C:\Users\123\AppData\Local\Temp\MSBuildTempFirstNameLastName(CompanyName` does exist. Could it be that there is an issue with the way MSBuild picks a temp directory?
 
My MSBuild version is as follows:
```powershell
PS C:\Users\123\source\repos> msbuild --version
MSBuild version 17.8.3+195e7f5a3 for .NET Framework
17.8.3.51904
```
I am using Windows 10 OS build 19044.3693.

I am not 100% if this is a CMake or MSBuild issue, but because the temporary path name starts with `MSBuildTemp`, and there was a [recent PR modifying the MSBuild temporary path directory](https://github.com/dotnet/msbuild/pull/9232), I suspect it may be related to the way MSBuild handles the temporary directories on Windows.

Many thanks in advance, and sorry if I missed providing required information. Please let me know if there is something missing.

### Steps to Reproduce

The issue seems to happen with any CMake project. I picked a random [super-simple CMake project](https://github.com/Danaozhong/Lightweight-C-ASCII-Table).

After cloning, run in a powershell:
```powershell
mkdir bin
cd bin
cmake ..
cmake --build .
```

### Expected Behavior

The `cmake --build .` should run successfully, and build the project.

### Actual Behavior

crash during building:
```powershell
PS C:\Work\CPP\Lightweight-C-ASCII-Table\bin> cmake --build .
MSBuild version 17.8.3+195e7f5a3 for .NET Framework

  'C:\Users\123\AppData\Local\Temp\MSBuildTempFirstNameLastName' is not recognized as an internal or external command,
  operable program or batch file.
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(237,5): error MSB8066: Custom build for 'C:\Wo
rk\CPP\Lightweight-C-ASCII-Table\bin\CMakeFiles\5d52f6b646fe4691cf3bf389522d7dd3\generate.stamp.rule' exited with code 1. [C:\Work\CPP\Lightweight-C-ASCII-T
able\bin\ZERO_CHECK.vcxproj]
```

**Comment by Danaozhong** [comment]
Upon closer look, it seems the issue was introduced with this commit: https://github.com/dotnet/msbuild/commit/7f3a30cd04aa884699b7980a9209aade28f1eca2#diff-3ecd811e3c6e9fd00cc8dd066953e6b280bedb594223203d0bb7786a46ca18d5L27

This PR seems to already fix the issue I observe: https://github.com/dotnet/msbuild/pull/9232

Let me try with a pre-release version, and see if the issue is fixed.

**Comment by Danaozhong** [comment]
I rebuilt `MSBuild` by using `v17.8.3` as a basis, and chery-pick #9232 on top of it. The rebuilt `MSBuild` now works without issues. I can therefore confirm that the issue raised here will be resolved with #9232.

**Comment by rokonec** [comment]
Closed as #9232 dupe
